File: backend/app/schemas/rental_document.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


DOCUMENT_TYPES = frozenset(
    {
        "lease_agreement",
        "inventory",
        "move_in_report",
        "payment_instructions",
        "house_rules",
        "other",
    }
)

DOCUMENT_TYPE_LABELS = {
    "lease_agreement": "Lease Agreement",
    "inventory": "Inventory",
    "move_in_report": "Move-in Report",
    "payment_instructions": "Payment Instructions",
    "house_rules": "House Rules",
    "other": "Other",
}


class RentalDocumentResponse(BaseModel):
    id: int
    viewing_request_id: int | None
    property_id: int
    renter_id: int
    realtor_id: int
    document_type: str
    display_name: str
    title: str | None
    original_filename: str | None
    mime_type: str
    size_bytes: int
    uploaded_by_user_id: int
    status: str
    can_download: bool
    can_archive: bool
    created_at: datetime
    updated_at: datetime
    archived_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class RentalDocumentListResponse(BaseModel):
    items: list[RentalDocumentResponse]
    total: int


class RentalDocumentUploadForm(BaseModel):
    document_type: Literal[
        "lease_agreement",
        "inventory",
        "move_in_report",
        "payment_instructions",
        "house_rules",
        "other",
    ]
    title: str | None = Field(default=None, max_length=200)

    @field_validator("title", mode="before")
    @classmethod
    def normalize_title(cls, value):
        if value is None:
            return None

        if isinstance(value, str):
            trimmed = value.strip()
            return trimmed or None

        return value

File: frontend/types/rentalDocument.ts
export type RentalDocumentType =
  | "lease_agreement"
  | "inventory"
  | "move_in_report"
  | "payment_instructions"
  | "house_rules"
  | "other"

export type RentalDocumentStatus = "active" | "archived"

export type RentalDocument = {
  id: number
  viewing_request_id: number | null
  property_id: number
  renter_id: number
  realtor_id: number
  document_type: RentalDocumentType
  display_name: string
  title: string | null
  original_filename: string | null
  mime_type: string
  size_bytes: number
  uploaded_by_user_id: number
  status: RentalDocumentStatus
  can_download: boolean
  can_archive: boolean
  created_at: string
  updated_at: string
  archived_at: string | null
}

export type RentalDocumentListResponse = {
  items: RentalDocument[]
  total: number
}

export const RENTAL_DOCUMENT_TYPE_OPTIONS: Array<{
  value: RentalDocumentType
  label: string
}> = [
  { value: "lease_agreement", label: "Lease Agreement" },
  { value: "inventory", label: "Inventory" },
  { value: "move_in_report", label: "Move-in Report" },
  { value: "payment_instructions", label: "Payment Instructions" },
  { value: "house_rules", label: "House Rules" },
  { value: "other", label: "Other" },
]

export function formatDocumentSize(bytes: number): string {
  if (bytes < 1024) {
    return `${bytes} B`
  }

  if (bytes < 1024 * 1024) {
    return `${(bytes / 1024).toFixed(1)} KB`
  }

  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`
}

export function getRentalDocumentStatusLabel(status: RentalDocumentStatus): string {
  if (status === "archived") {
    return "Archived"
  }

  return "Active"
}
